### PR TITLE
BsonType configuration 

### PR DESCRIPTION
**Pull request by damieng**
Provides a `HasBsonType` fluent API that allows the model builder to configure specific properties to be stored as specific BSON types.

Fixes EF-127

A subsequent PR will further expose this feature out so that `BsonRepresentation` attributes can be recognized (EF-44)

### CODE REVIEW

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs [hunk 1]
```diff
@@ -53,65 +54,127 @@ public static class MongoPropertyBuilderExtensions
         => (PropertyBuilder<TProperty>)HasElementName((PropertyBuilder)propertyBuilder, name);
 
     /// <summary>
-    /// Configures the <see cref="DateTimeKind"/> for the property.
+    /// Configures the document element that the property is mapped to when targeting MongoDB.
+    /// If an empty string is supplied then the property will not be persisted.
```
Comment (JamesKovacs) [inline src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:58]: Although not recommended, an empty string is a valid field name in MongoDB. `null` however is not a valid field name. Is an empty string not mapping a property an EF convention or something that we are introducing?

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs [hunk 2]
```diff
@@ -53,65 +54,127 @@ public static class MongoPropertyBuilderExtensions
...
+    /// <returns>The same builder instance if the configuration was applied, <see langword="null" /> otherwise.</returns>
+    public static IConventionPropertyBuilder? HasElementName(
+        this IConventionPropertyBuilder propertyBuilder,
+        string? name,
+        bool fromDataAnnotation = false)
     {
-        propertyBuilder.Metadata.SetDateTimeKind(dateTimeKind);
+        if (!CanSetElementName(propertyBuilder, name, fromDataAnnotation))
+        {
+            return null;
```
Comment (JamesKovacs) [inline src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:71]: Should we not be throwing an `InvalidOperationException` in this case? If we return `null`, then we will throw a `NullReferenceException` when we chain the fluent API.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs [hunk 3]
```diff
@@ -53,65 +54,127 @@ public static class MongoPropertyBuilderExtensions
...
+    /// </summary>
+    /// <param name="propertyBuilder">The builder for the property being configured.</param>
+    /// <param name="bsonType">The <see cref="BsonType"/> to store this property as.</param>
+    /// <returns>The same builder instance so that multiple calls can be chained.</returns>
+    public static PropertyBuilder HasBsonType(
+        this PropertyBuilder propertyBuilder,
+        BsonType bsonType)
     {
-        propertyBuilder.Metadata.SetDateTimeKind(dateTimeKind);
+        ArgumentNullException.ThrowIfNull(bsonType);
```
Comment (JamesKovacs) [inline src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:101]: `BsonType` is an enum and cannot be null.

Comment (JamesKovacs) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:101]: Do we want to make this `BsonType?` so that `null` is valid and can be used to unset a previous `HasBsonType` configuration?

File: src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs [hunk 4]
```diff
@@ -103,19 +113,26 @@ private static IBsonSerializer CreateTypeSerializer(Type type, IReadOnlyProperty
...
-            _ when type == typeof(Decimal128) => new Decimal128Serializer(),
             _ when type.IsEnum => EnumSerializer.Create(type),
+
+            // Nullable
             {IsGenericType: true} when type.GetGenericTypeDefinition() == typeof(Nullable<>)
                 => CreateNullableSerializer(type.GetGenericArguments()[0]),
+
+            // Mongo structs
+            _ when type == typeof(ObjectId) => ObjectIdSerializer.Instance,
+            _ when type == typeof(Decimal128) => new Decimal128Serializer(),
```
Comment (JamesKovacs) [inline src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs:131]: We should configure serializers for ObjectId and Decimal128, but this seems unrelated to the PR at hand. Move to a separate PR?

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Exerciser.cs
Comment (JamesKovacs) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Exerciser.cs:1]: This seems related to adding support for `ObjectIdSerializer`, not `HasBsonType` support.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Utilities/ExtensionMethods.cs [hunk 5]
```diff
@@ -36,9 +37,12 @@ public static byte NextByte(this Random random)
...
-        =>  dateTime.ToString("yyyy-MM-ddTHH:mm:ss.fffK");
-
     public static void WriteTestDocs<T>(this IMongoCollection<T> collection, IEnumerable<T> docs)
         => collection.BulkWrite(docs.Select(p => new InsertOneModel<T>(p)));
+
+    public static DateTime ToBsonPrecision(this DateTime dateTime)
+    {
+        var bsonDateTime = new BsonDateTime(dateTime);
+        return dateTime.Kind == DateTimeKind.Utc ? bsonDateTime.ToUniversalTime() : bsonDateTime.ToLocalTime();
+    }
```
Comment (JamesKovacs) [inline tests/MongoDB.EntityFrameworkCore.FunctionalTests/Utilities/ExtensionMethods.cs:47]: Unrelated to `HasBsonType` support. Move to separate PR.

**Review by JamesKovacs — CHANGES_REQUESTED** [review]
`HasBsonType` looks good but the unrelated work should be moved into their own PRs.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs (hunk 2 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:71]: This is the pattern used across all EF providers. You can see it in the Cosmos, SqlServer and relational base providers.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs (hunk 1 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:58]: This was just improving the documentation. EF uses this convention in other providers. If we want to support writing to an element with an empty name we may be able to that but we'd need to create a new ticket and figure out an alternative marker for unmapped properties.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs (hunk 3 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs:101]: Yeah, let's do that.

File: src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs (hunk 4 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs:131]: This doesn't add/remove any support, it just groups them together with comments as the list was getting long. I can revert.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Exerciser.cs
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/Exerciser.cs:1]: Just a minor reordering while I was in there. No new code.

Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/Exerciser.cs:1]: This is a new class for testing round-tripping data back and forth to MongoDB with an extra check via the C# driver. It came out of writing all the new BsonType tests as they were getting repetitive. It's not related to any other work.

File: tests/MongoDB.EntityFrameworkCore.FunctionalTests/Utilities/ExtensionMethods.cs (hunk 5 above)
Comment (damieng) [reply tests/MongoDB.EntityFrameworkCore.FunctionalTests/Utilities/ExtensionMethods.cs:47]: This is test-only code and relates to testing DateTime's which I need for the new BsonType tests that touch BsonDateTime. 

I had a ToExpectedPrecision method I was previously using for some other BsonDateTime related tests but it didn't handle UTC. The new method is used for both the new tests and the old.

File: src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs (hunk 3 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Extensions/MongoPropertyBuilderExtensions.cs]: Now done.

File: src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs (hunk 4 above)
Comment (damieng) [reply src/MongoDB.EntityFrameworkCore/Serializers/SerializationHelper.cs]: Removed this clean-up from the PR.

**Review by JamesKovacs — APPROVED** [review]
LGTM
